### ISSUE TITLE
standardize all stringification API's around binary $ operator + variadic strAppend

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/RFCs#191]
API's like `$`(a: MyType) (which are what we almost exclusively use in stdlib, nimble etc) do not compose efficiently because they create temporaries, allocate etc.
instead, we should build API's around appending. This gives best of both worlds:
* efficiency where needed
* still possible to use old way `$a` wherever it's convenient, thanks a single unary `$` overload:
```nim
template `$`*[T](a: T): string =
  ## the only unary `$` that's ever needed
  var result: string
  result$a
  result
```

## note
`outplace` operator is not relevant to this discussion, we don't want to use outplace operator for something so common like stringification

## proposal: part 1
the API to stringify `MyType` shall be:
```nim
proc `$`(result: var string, a: MyType)
```

## example
```nim
import system except `$`

## overloads of binary `$` for all types we need to support
proc `$`*(result: var string, a: float) = result.addFloat a
proc `$`*(result: var string, a: int) = result.addInt a
template `$`*(result: var string, a: string) = result.add a

template `$`*[T](a: T): string =
  ## the only unary `$` that's ever needed
  var result: string
  result$a
  result
```

## proposal: part 2
variadic `strAppend` is always preferable to `&` + `$`, and often preferable to `varargs[string,$]`

following macro `strAppend` advantageously replaces many patterns, yet is as efficient as can be.
In particular, it allows writing a more efficient `echo(a: varargs[string, $])` and `write(a: File, args: varargs[string, $])`, without introducing a bunch of temporaries.

```nim
import macros

macro strAppend*(dst: var string, args: varargs[untyped]): void =
  result = newStmtList()
  for ai in args:
    result.add quote do:
      `dst`$`ai`

macro strAppend*(args: varargs[untyped]): string =
  result = newStmtList()
  let dst = genSym(nskVar, "dst")
  result.add quote do:
    var `dst`: string
  for ai in args:
    result.add quote do:
      `dst`$`ai`
  result.add quote do:
    `dst`
```

## usage
```nim
template echo*(args: varargs[untyped]) =
  ## more efficient than system.echo: no temporaries
  ## could be further optimized using a fixed size buffer when short enough or even a threadvar string to do all the appending
  write(stdout, strAppend(args, "\n"))

proc main()=
  var x = 0.3
  var ret: string
  ret$x
  ret$" bar "
  ret$12
  doAssert ret == "0.3 bar 12"
  doAssert $0.3 == "0.3"

  var ret2: string
  ret2.strAppend 0.1, " foo ", 12
  doAssert ret2 == "0.1 foo 12"
  doAssert strAppend(0.1, " foo ", 12) == "0.1 foo 12"
  echo 0.1, " foo ", 12 # prints `0.1 foo 12`

main()
```

**Comment by krux02** [comment]
Yes, I agree. I want this. And we almost already have it. [Please take a look at ``strformat.formatValue``](https://github.com/nim-lang/Nim/blob/240174dd81781e279d84090f1937ebf4b971e29c/lib/pure/strformat.nim#L420). It even allows an [extendable format specifier](https://github.com/nim-lang/Nim/blob/240174dd81781e279d84090f1937ebf4b971e29c/tests/stdlib/tstrformat.nim#L90) and does not have intermediate strings (unless of course the [inevitable `$` fallback](https://github.com/nim-lang/Nim/blob/240174dd81781e279d84090f1937ebf4b971e29c/lib/pure/strformat.nim#L515) triggers secretly).

**Comment by Araq** [comment]
Super nice RFC, one minor correction, ``strAppend`` should be called ``concat`` IMHO.

**Comment by Clyybber** [comment]
@disruptek Had the awesome idea to transform `x = a & b` into `a.add b; x = a`.
We can do that whenever a isn't used afterwards (even better when b isn't used afterwards either, then we can sink it).

**Comment by Varriount** [comment]
I generally agree with this. I'm a bit hesitant about the new usage of the `$` operator, but it makes sense.

**Comment by Araq** [comment]
It was noticed on IRC that we already have the binary append operator and it's written as ``&=``. It doesn't do custom tostringifcations though. In general people dislike new operators so maybe we should stick to ``proc toString(res: var string; x: CustomType)`` which is currently unused and so doesn't conflict with anything.

**Comment by Varriount** [comment]
Perhaps "write" or "writeString" would be better?

**Comment by krux02** [comment]
> I have a WIP that would allow writing strAppend as a template instead of a macro, to avoid depending on macros.nim

Well, what I have in mind for quite some time now is the ability to forward varargs in a template without importing macros. That would also be really helpful to implement ``bitops`` without macros. And `bitops` without macros would mean it could be used in ``sets.nim``.

> ``## more efficient than system.echo: no temporaries``

That is not correct. There is still one temporary per call to echo.
In https://github.com/nim-lang/Nim/pull/13277 I have a concept that really has no temporary.
